core/gekkoStream.js

```diff
@@ -1,30 +1,84 @@
 // Small writable stream wrapper that
 // passes data to all `candleConsumers`.
 
-var Writable = require('stream').Writable;
-var _ = require('lodash');
-var async = require('async');
+const Writable = require('stream').Writable;
+const _ = require('lodash');
+const async = require('async');
+const moment = require('moment');
 
-var util = require('./util');
-var env = util.gekkoEnv();
-var mode = util.gekkoMode();
+const util = require('./util');
+const env = util.gekkoEnv();
+const mode = util.gekkoMode();
+const config = util.getConfig();
+const log = require(util.dirs().core + 'log');
 
-var Gekko = function(candleConsumers) {
-  this.candleConsumers = candleConsumers;
+var Gekko = function(plugins) {
+  this.plugins = plugins;
+  this.candleConsumers = plugins
+    .filter(plugin => plugin.processCandle);
   Writable.call(this, {objectMode: true});
 
+  this.producers = this.plugins
+    .filter(p => p.meta.emits);
+
   this.finalize = _.bind(this.finalize, this);
 }
 
 Gekko.prototype = Object.create(Writable.prototype, {
   constructor: { value: Gekko }
 });
 
-Gekko.prototype._write = function(chunk, encoding, _done) {
-  var done = _.after(this.candleConsumers.length, _done);
-  _.each(this.candleConsumers, function(c) {
-    c.processCandle(chunk, done);
-  });
+if(config.debug) {
+  // decorate with more debug information
+  Gekko.prototype._write = function(chunk, encoding, _done) {
+
+    const start = moment();
+    var relayed = false;
+    var at = null;
+
+    const timer = setTimeout(() => {
+      if(!relayed)
+        log.error([
+          `The plugin "${at}" has not processed a candle for 1 second.`,
+          `This will cause Gekko to slow down or stop working completely.`
+        ].join(' '));
+    }, 1000);
+
+    const flushEvents = _.after(this.candleConsumers.length, () => {
+      relayed = true;
+      clearInterval(timer);
+      this.flushDefferedEvents();
+      _done();
+    });
+    _.each(this.candleConsumers, function(c) {
+      at = c.meta.name;
+      c.processCandle(chunk, flushEvents);
+    }, this);
```

Comment:
the `context` argument, that existed in underscore, doesn't seem to exist in lodash